c_src/sqlite3_nif.c

```diff
@@ -691,6 +691,120 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+    ERL_NIF_TERM head, tail;
+
+    // process types
+    for (int i = 0; i < stmt_param_count; i++) {
+        if (!enif_get_list_cell(env, types, &head, &tail)) {
+            return raise_badarg(env, types);
+        }
+
+        int type;
+        if (!enif_get_int(env, head, &type)) {
+            return raise_badarg(env, types);
+        }
+
+        types_array[i] = type;
+        types          = tail;
+    }
+
+    int rc;
+
+    // process rows
+    while (enif_get_list_cell(env, rows, &head, &tail)) {
+        sqlite3_reset(statement->statement);
+
+        // bind row
+        for (int i = 1; i <= stmt_param_count; i++) {
+            ERL_NIF_TERM param;
+
+            if (!enif_get_list_cell(env, head, &param, &head)) {
+                return raise_badarg(env, param);
+            }
+
+            if (enif_is_identical(param, am_nil)) {
+                rc = sqlite3_bind_null(statement->statement, i);
+            } else {
+                switch (types_array[i - 1]) {
+                    case SQLITE_INTEGER: {
+                        ErlNifSInt64 i64;
+                        if (!enif_get_int64(env, param, &i64)) {
+                            return raise_badarg(env, param);
+                        }
+
+                        rc = sqlite3_bind_int64(statement->statement, i, i64);
+                        break;
+                    }
+
+                    case SQLITE_FLOAT: {
+                        double f64;
+                        if (!enif_get_double(env, param, &f64)) {
+                            return raise_badarg(env, param);
+                        }
+
+                        rc = sqlite3_bind_double(statement->statement, i, f64);
+                        break;
+                    }
+
+                    case SQLITE_TEXT: {
+                        ErlNifBinary text;
+                        if (!enif_inspect_binary(env, param, &text)) {
+                            return raise_badarg(env, param);
+                        }
+
+                        rc = sqlite3_bind_text(statement->statement, i, (char*)text.data, text.size, SQLITE_TRANSIENT);
+                        break;
+                    }
+
+                    case SQLITE_BLOB: {
+                        ErlNifBinary blob;
+                        if (!enif_inspect_binary(env, param, &blob)) {
+                            return raise_badarg(env, param);
+                        }
+
+                        rc = sqlite3_bind_blob(statement->statement, i, (char*)blob.data, blob.size, SQLITE_TRANSIENT);
+                        break;
+                    }
+                }
+            }
+
+            if (rc != SQLITE_OK) {
+                return make_sqlite3_error_tuple(env, rc, sqlite3_db_handle(statement->statement));
+            }
+        }
+
+        rc = sqlite3_step(statement->statement);
+        if (rc != SQLITE_DONE) {
+            return make_sqlite3_error_tuple(env, rc, sqlite3_db_handle(statement->statement));
```

Comment:
We are missing cleanup code here. We need to `enif_free(types_array)` before all of these error returns.

Reply:
Ah, right.

Reply:
I think I'll need to find another way to pass the types information. The current approach is getting complicated.

Reply:
Sorry!

Reply:
> This is just the nature of C.

I think it should be possible to implement `insert_all` or a similar function for bulk inserts with no manual memory allocation. It's only used in a small part of the function -- just to know what types to insert -- but ends up affecting all control flow.

Reply:
All `Repo.insert_all` calls would go through this NIF and the performance improvement is quite significant (seems close to x2 from my benchmarks, but potentially higher becuase it allows reusing a single prepared statement for all inserts into a single table) so the I think it's worth it.